pkg/persistence/postgrespersister.go

```diff
@@ -115,19 +142,104 @@ func (p *PostgresPersister) PopulateBlockDataFromDB(tableName string) error {
+	// Create version table first if it dne
+	versionTableQuery := postgres.CreateVersionTableQuery()
+	_, err := p.db.Exec(versionTableQuery)
+	if err != nil {
+		return err
+	}
+	if version == "" {
+		version, err = p.RetrieveVersion()
+		if err != nil {
+			if err == cpersist.ErrPersisterNoResults {
+				return fmt.Errorf("No version in version table, specify a version: err %v", err)
+			}
+			return err
+		}
+	}
+	p.version = version
+	err = p.SaveVersion(p.version)
+	if err != nil {
+		return err
+	}
+	eventTableName := p.GetTableName(postgres.EventTableType)
+	eventTableQuery := postgres.CreateEventTableQuery(eventTableName)
+	_, err = p.db.Exec(eventTableQuery)
 	return err
 }
 
 // CreateIndices creates the indices for DB if they don't exist
 func (p *PostgresPersister) CreateIndices() error {
-	indexQuery := postgres.CreateEventTableIndices()
+	eventTableName := p.GetTableName(postgres.EventTableType)
+	indexQuery := postgres.CreateEventTableIndices(eventTableName)
 	_, err := p.db.Exec(indexQuery)
 	return err
 }
 
+func (p *PostgresPersister) updateDBQueryBuffer(updatedFields []string, tableName string, dbModelStruct interface{}) (bytes.Buffer, error) {
+	var queryBuf bytes.Buffer
+	queryBuf.WriteString("UPDATE ") // nolint: gosec
+	queryBuf.WriteString(tableName) // nolint: gosec
+	queryBuf.WriteString(" SET ")   // nolint: gosec
+	for idx, field := range updatedFields {
+		dbFieldName, err := cpostgres.DbFieldNameFromModelName(dbModelStruct, field)
+		if err != nil {
+			return queryBuf, fmt.Errorf("Error getting %s from %s table DB struct tag: %v", field, tableName, err)
+		}
+		queryBuf.WriteString(fmt.Sprintf("%s=:%s", dbFieldName, dbFieldName)) // nolint: gosec
+		if idx+1 < len(updatedFields) {
+			queryBuf.WriteString(", ") // nolint: gosec
+		}
+	}
+	return queryBuf, nil
+}
+
+func (p *PostgresPersister) retrieveVersionFromTable(tableName string) (string, error) {
+	dbVersionData := []postgres.VersionData{}
+	queryString := fmt.Sprintf(`SELECT * FROM %s WHERE service_name=$1;`, tableName) // nolint: gosec
+	err := p.db.Select(&dbVersionData, queryString, crawlerServiceName)
+	if err != nil {
+		return "", err
+	}
+	if len(dbVersionData) == 0 {
+		return "", cpersist.ErrPersisterNoResults
+	}
+	if len(dbVersionData) > 1 {
+		return "", fmt.Errorf("There shouldn't be more than one version type in DB for %v", crawlerServiceName)
+	}
+	return dbVersionData[0].Version, nil
+}
+
+// saveVersionToTable saves/updates the version
+func (p *PostgresPersister) saveVersionToTable(tableName string, versionNumber string) error {
+	dbVersionStruct := postgres.VersionData{
+		Version:     versionNumber,
+		ServiceName: crawlerServiceName}
+
+	queryString, err := p.updateDBQueryBuffer([]string{versionFieldName}, tableName, dbVersionStruct)
+	if err != nil {
+		return fmt.Errorf("Error creating query string %v", err)
+	}
+	queryString.WriteString(" WHERE service_name=:service_name;") // nolint: gosec
```

Comment:
Seems like a perfect place for use an upsert: http://www.postgresqltutorial.com/postgresql-upsert/  I think I may have added a postgresl upsert somewhere in the code, i'll see if i can dig that up.

some more info in the postgres docs by searching for upsert: https://www.postgresql.org/docs/9.6/sql-insert.html

Reply:
Also, if we just overwrite the existing version entry for a service name, what is the plan to clean out the old tables?  Are we going to go through through all the valid table names and find the ones that do not have this version in the name?

Reply:
Thank you! Will change to upsert. 
Yes, that was the plan. But...we  could also save all versions, and retrieve for the latest version on each call to `RetrieveVersion`. Then the service that deletes tables will retrieve all older versions, delete those rows and tables

Reply:
yeah, that might work. i don't want to overcomplicate things, but a nice side effect of keeping all the versions and when they deployed is that we have a record of them all.